Core multiple stacks management

yufangzhang opened the pull request:
- Modify fab task cfn_create, cfn_delete to mange multiple stacks.
- Create task set_active_stack(tag) to switch among stacks
- Add unit tests for new methods

Commit:
Add parallel stack support

update cfn_create + cfn_delete with multiple stacks support
add set_active_stack fab task

mattpep commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -393,39 +397,90 @@ def set_stack_name():
+    """
+    r53_conn = get_connection(R53)
+    zone_id = get_zone_id()
+    record = "{}".format(get_tag_record_name(stack_tag))
```
This could probably be simply:

``` python
record = get_tag_record_name(stack_tag)
```

i.e. without the call to `format`.

mattpep commented on bootstrap_cfn/errors.py:
```diff
@@ -77,3 +77,33 @@ def __init__(self, autoscaling_group, expected_instance_count, instances):
+
+class ActiveTagExistConflictError(BootstrapCfnError):
+    def __init__(self, stack_id):
+        msg = ("An active record already exists in. ".format(stack_id))
```
The `format` parameter is not used. Presumably something like the following is intended:

``` python
msg = 'An active record already exists in {}.'.format(stack_id)
```

mattpep commented on bootstrap_cfn/r53.py:
```diff
@@ -68,14 +69,71 @@ def update_dns_record(self, zone, record, record_type, record_value, is_alias=Fa
+            record_name(String):
+            record_type:
+        Returns:
+            String
```
I suggest adding the following text to this line:

`or None, in the event of there being no A or TXT record`

mattpep commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +802,118 @@ def cycle_instances(delay=None):
+
+def get_all_elbs():
+    """
+    Returns all available elbs from cloudformation configuration
```
The docstring should be a bit more clear in its description -e.g.

`Returns all internet-facing elbs from cloudformation configuration`

sevenmachines commented:
@yufangzhang happy to merge after those couple of minor things @mattpep mentioned are fixedup!

yufangzhang commented:
@mattpep I've done the changes. can you have a look ~?

sevenmachines commented:
@mattpep @yufangzhang we happy to merge?

mattpep commented:
LGTM. Ideally we should squash the fixup commit into the one it updates. Then it's good to merge.

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -393,39 +397,90 @@ def set_stack_name():
+    Args:
+        stack_tag: the tag of stack
+    Returns:
+        True if stack exists
```
Is this accurate? To be specific does it return "True" if stack exists?

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -483,35 +540,80 @@ def cfn_delete(force=False, pre_delete_callbacks=None):
+    if not force:
+        x = raw_input("Are you really sure you want to blow away the whole stack for {}!? (y/n)\n".format(stack_name))
+        if x not in ['y', 'Y', 'Yes', 'yes']:
+            sys.exit(1)
```
Why did you move this?

yufangzhang replied:
because when it comes to "deleting active stack", I want to delete "active" dns records while keep the stack. 
The reason I put cfn.delete after dns deleted is though, previously it stucked in vpc deleting. I wanna put it at the end so that I can killed the program and delete vpc manually in AWS console if neccessary.

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -483,35 +540,80 @@ def cfn_delete(force=False, pre_delete_callbacks=None):
+            get_zone_name(), get_zone_id(), tag_record_name, 'TXT'))
+        record_name = '{}.{}'.format(tag_record_name, get_zone_name())
+        if stack_id and stack_id == record_value[1:-1]:
+            r53_conn.delete_dns_record(get_zone_id(), record_name, 'TXT', record_value)
```
In this method you seem to be using get_zone_id() about 9 times. Does this get cached by the boto library? We shouldn't be repeating calls pointlessly as there is a risk to get throttled. 
Also I had created `get_cached_zone_id` that will do some caching when it looks up zone ids. Actually the whole get_zone_id() can be replicated by refactoring to this and passing an empty cache dict if you don't want caching: 

``` python
def get_zone_id(r53_conn):
    return get_cached_zone_id(r53_conn, {}, get_zone_name())
```

yufangzhang replied:
actually they're the same id...I'll remove the duplicated calls.

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +802,118 @@ def cycle_instances(delay=None):
+                                   '"{}"'.format(tag_stack_id))
+        logger.info("fab_tasks::set_active_stack: Successfully updated dns alias record")
+    except Exception as e:
+        logger.error("fab_tasks::set_stack_name:"
```
`set_active_stack` instead of `set_stack_name`

yufangzhang replied:
a yeah

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +802,118 @@ def cycle_instances(delay=None):
+            logger.info("fab_tasks::set_active_stack: Successfully updated dns alias record")
+            print green("Active stack is switched to {}".format(tag_record))
+        except Exception as e:
+            logger.error("fab_tasks::set_stack_name:"
```
`set_active_stack` instead of `set_stack_name`

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +802,118 @@ def cycle_instances(delay=None):
+        logger.error("fab_tasks::set_stack_name:"
+                     "Error updating dns record: "
+                     "{}".format(e.message))
+
```
Is there supposed to be a `raise` here? If the record fails to get updated?

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +802,118 @@ def cycle_instances(delay=None):
+                     "{}".format(e.message))
+
+    # get all public facing elbs
+    elb = get_one_public_elbs()
```
comment misleading. you are only getting the first.

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +802,118 @@ def cycle_instances(delay=None):
+    return elbs
+
+
+def get_one_public_elbs():
```
consider renaming to `get_first_public_elb` or something more appropriate.

mattpep replied:
and if there is more than one public ELB, how are they ordered? In other words, how can one know which will be the first, and whether it will be the correct one?

filipposc5 commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +802,118 @@ def cycle_instances(delay=None):
+    """
+    elbs = get_all_elbs()
+    if len(elbs) < 1:
+        raise PublicELBNotFoundError
```
Is this really an invalid case, not having a public ELB?

yufangzhang replied:
currently we find internet facing elbs from cloudformation so if we don't get any, should check its configuration file...I'll specify it in the error message of PublicELBNotFoundError

filipposc5 replied:
All I am saying is that there are two cases where we can create a stacks that don't fit this restriction you have imposed. 1) A stack without an ELB 2) a stack with a non internet facing ELB. I haven't tested it but at least the 2nd case is valid I believe. So maybe raising an exception is not suitable here.

filipposc5 commented on bootstrap_cfn/config.py:
```diff
@@ -51,7 +51,7 @@ class ConfigParser(object):
         self.stack_name = stack_name
         self.data = data
-
+        self.stack_id = self.stack_name.split('-')[-1]
```
how can you have access to the stack id here? since this generates the template to create the stack id to begin with.  You might be able to use  `AWS::StackId` (or maybe `AWS::StackName` ) with cloudformation syntax to obtain the ID.

yufangzhang replied:
ah we created stack_name in the format of [app-env-stackid] so can access it...

filipposc5 replied:
Disregard this comment of mine, as we saw this is fine.

yufangzhang commented:
@mattpep yeup, updated!

mattpep commented:
The first bullet point of this PR makes reference to changing the task `cfn_create` but I see no changes of that task within `config.py`.
Does the stack_id now have to be specified? More generically, it'd be useful for the documentation be updated to explain how these new tasks work.

yufangzhang commented:
@mattpep we generate `stack_id` from `uuid` in `set_stack_name` at the beginning as what it used to be, then add it to `stack_name`. In config.py, we only added `self.stack_id` as instance variable in config and changed the format of dns record name.

filipposc5 commented on bootstrap_cfn/r53.py:
```diff
@@ -68,14 +69,71 @@ def update_dns_record(self, zone, record, record_type, record_value, is_alias=Fa
                 if rr.type == 'TXT':
                     rr.resource_records[0] = rr.resource_records[0][1:-1]
+                if rr.type == 'A':
+                    return rr.alias_dns_name
```
An A record doesn't have to be a DNS alias.

yufangzhang replied:
yes. But we wanted to keep returns as String type instead of "either String( `TXT` ) or RecordObejct ( `A` )".

Commit:
Add unittest for multiple stacks

Add unittest on newly added methods.

Commit:
Update flake8-import-order to 2a595f1

Test starts failing with the error below, which is fixed in 2a595f1
(aka version 0.9 unreleased):
```
Traceback (most recent call last):
  File "/Users/yufang/.virtualenvs/test/bin/flake8", line 11, in <module>
    sys.exit(main())
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/main/cli.py", line 16, in main
    app.run(argv)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/main/application.py", line 316, in run
    self._run(argv)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/main/application.py", line 299, in _run
    self.initialize(argv)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/main/application.py", line 290, in initialize
    self.register_plugin_options()
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/main/application.py", line 150, in register_plugin_options
    self.check_plugins.register_options(self.option_manager)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/plugins/manager.py", line 451, in register_options
    list(self.manager.map(register_and_enable))
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/plugins/manager.py", line 261, in map
    yield func(self.plugins[name], *args, **kwargs)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/plugins/manager.py", line 447, in register_and_enable
    call_register_options(plugin)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/plugins/manager.py", line 357, in generated_function
    return method(optmanager, *args, **kwargs)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8/plugins/manager.py", line 207, in register_options
    add_options(optmanager)
  File "/Users/yufang/.virtualenvs/test/lib/python2.7/site-packages/flake8_import_order/flake8_linter.py", line 32, in add_options
    parser.config_options.append("application-import-names")
AttributeError: 'OptionManager' object has no attribute 'config_options'
```

cliffxuan commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -393,39 +396,70 @@ def set_stack_name():
-    r53_conn = get_connection(R53)
-    zone_name = cfn_config.data.get('master_zone', None)
+    record = "{}.{}".format(get_tag_record_name(stack_tag), zone_name)
+    logger.info("fab_tasks::set_stack_name: "
```
`format` is better than using `%s` in formatting a string, however, in `logging`, `%s` is actually preferred, so you would do:

```
logger.info("Creating stack suffix '%s' for record '%s'... %s...", stack_suffix, record, zone_id)
```

cliffxuan commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -393,39 +396,70 @@ def set_stack_name():
+    r53_conn = get_connection(R53)
+    zone_name = get_zone_name()
+    zone_id = get_zone_id()
+    stack_suffix = uuid.uuid4().__str__()[-8:]
```
you can do `uuid.uuid4().hex[-8:]`

cliffxuan commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -393,39 +396,70 @@ def set_stack_name():
+    zone_name = get_zone_name()
+    zone_id = get_zone_id()
+    stack_suffix = uuid.uuid4().__str__()[-8:]
     if hasattr(env, 'tag'):
```
EAFP is a more pythonic way

```
try:
      if env.tag == 'active':
      ....
except AttributeError:
    env.tag = stack_suffix 
```

cliffxuan commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -393,39 +396,70 @@ def set_stack_name():
+    try:
+        r53_conn.update_dns_record(zone_id, record, 'TXT', '"{0}"'.format(stack_suffix))
+        env.stack_name = "{0}-{1}".format(get_legacy_name(), stack_suffix)
+    except:
```
always use `except Exception:` instead of just `except:` to trap exceptions in general. `except` traps other things like `KeyboardInterrupt`.

cliffxuan commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -393,39 +396,70 @@ def set_stack_name():
+
+
+def get_zone_name():
+    zone_name = get_basic_config().get('master_zone', None)
```
better do the EAFP way:

```
try:
    zone_name = get_basic_config()['master_zone']
except KeyError:
    raise  CfnConfigError("No master_zone in yaml, unable to create/find DNS records for stack name")
```

cliffxuan commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -701,3 +763,114 @@ def cycle_instances(delay=None):
+        main_record_value = r53_conn.get_record(zone_name, zone_id, elb, 'A')
+    except:
+        print green("No active stack exists.")
+        return None
```
could just do `return`, which is the same as `return None`

cliffxuan commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -18,7 +15,10 @@
 from bootstrap_cfn.config import ConfigParser, ProjectConfig
 from bootstrap_cfn.elb import ELB
-from bootstrap_cfn.errors import BootstrapCfnError, CfnConfigError, CloudResourceNotFoundError, DNSRecordNotFoundError, ZoneIDNotFoundError
+from bootstrap_cfn.errors import ActiveTagExistConflictError, BootstrapCfnError,\
```
preferred way to write long import line is to use `()` rather than `\` https://www.python.org/dev/peps/pep-0328/

```
from bootstrap_cfn.errors import (ActiveTagExistConflictError, BootstrapCfnError,
    CfnConfigError, CloudResourceNotFoundError, DNSRecordNotFoundError,
    ...)
```

yufangzhang commented:
@cliffxuan Thanks for the advice Cliff! I've made some changes to make it more pythonic 👻 
@lukaszraczylo think you may have used this to do the CCCD stuff? I've just updated a bug on deleting active stack records. hopefully you haven't come across that....
@filipposc5 @niallcreech I think this beast is ready to be merged, can anyone have a look? sorry about its size :P
